File: src/main/java/net/hectus/neobb/turn/default_game/block/TSponge.java
package net.hectus.neobb.turn.default_game.block;

import net.hectus.neobb.buff.Buff;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.turn.default_game.CounterFilter;
import net.hectus.neobb.turn.default_game.attributes.clazz.NeutralClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.SupernaturalClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.WaterClazz;
import net.hectus.neobb.turn.default_game.attributes.function.CounterbuffFunction;
import org.bukkit.block.Block;
import org.bukkit.potion.PotionEffectType;

import java.util.List;

public class TSponge extends BlockTurn implements CounterbuffFunction, WaterClazz {
    public TSponge(NeoPlayer player) { super(player); }
    public TSponge(Block data, NeoPlayer player) { super(data, player); }

    @Override
    public boolean requiresUsageGuide() { return true; }

    @Override
    public int cost() {
        return 5;
    }

    @Override
    public void apply() {
        if (player.game.history().isEmpty()) return;
        if (player.game.history().getLast() instanceof TWater) {
            player.game.world().setStorm(true);
            new Buff.Luck(Buff.BuffTarget.YOU, 10).apply(player);
            new Buff.Effect(Buff.BuffTarget.YOU, PotionEffectType.JUMP_BOOST).apply(player);
        }
    }

    @Override
    public List<Buff> buffs() {
        return List.of(new Buff.ExtraTurn(), new Buff.Luck(Buff.BuffTarget.YOU, 5));
    }

    @Override
    public List<CounterFilter> counters() {
        return List.of(CounterFilter.clazz(NeutralClazz.class), CounterFilter.clazz(WaterClazz.class), CounterFilter.clazz(SupernaturalClazz.class));
    }
}

File: src/main/kotlin/net/hectus/neobb/turn/default_game/block/TSponge.kt
package net.hectus.neobb.turn.default_game.block

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.buff.Buff
import net.hectus.neobb.buff.Effect
import net.hectus.neobb.buff.ExtraTurn
import net.hectus.neobb.buff.Luck
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.turn.default_game.CounterFilter
import net.hectus.neobb.turn.default_game.attribute.clazz.NeutralClazz
import net.hectus.neobb.turn.default_game.attribute.clazz.SupernaturalClazz
import net.hectus.neobb.turn.default_game.attribute.clazz.WaterClazz
import net.hectus.neobb.turn.default_game.attribute.function.CounterbuffFunction
import org.bukkit.block.Block
import org.bukkit.potion.PotionEffectType

class TSponge(data: Block?, cord: Cord?, player: NeoPlayer?) : BlockTurn(data, cord, player), CounterbuffFunction, WaterClazz {
    override val cost: Int = 5
    override val requiresUsageGuide: Boolean = true

    override fun apply() {
        if (player!!.game.history.isEmpty()) return

        if (player.game.history.last() is TWater) {
            player.game.world.setStorm(true)
            Luck(10).apply(player)
            Effect(PotionEffectType.JUMP_BOOST).apply(player)
        }
    }

    override fun counters(): List<CounterFilter> {
        return listOf(CounterFilter.clazz(NeutralClazz::class), CounterFilter.clazz(WaterClazz::class), CounterFilter.clazz(SupernaturalClazz::class))
    }

    override fun buffs(): List<Buff<*>> {
        return listOf(ExtraTurn(), Luck(5))
    }
}